chaste_codegen/be_model.py

```diff
@@ -0,0 +1,314 @@
+            else:
+                result = BeModel.KINDS.Nonlinear
+        return result
+
+    def _get_non_linear_state_vars(self):
+
+        # return the state var part from the derivative equations where the rhs is not linear
+        return sorted([eq.lhs.args[0] for eq in self._derivative_equations
+                       if isinstance(eq.lhs, sp.Derivative) and
+                       eq.lhs.args[0] != self._membrane_voltage_var and
+                       self._check_expr(eq.rhs, eq.lhs.args[0]) != BeModel.KINDS.Linear],
+                      key=lambda s: self._printer.doprint(s))
+
+    def _format_rearranged_linear_derivs(self):
+        def rearrange_expr(expr, var):  # expr already in piecewise_fold form
+            """Rearrange an expression into the form g + h*var."""
+            if isinstance(expr, sp.Piecewise):
+                # The tests have to move into both components of gh:
+                # "if C1 then (a1,b1) elif C2 then (a2,b2) else (a0,b0)"
+                # maps to "(if C1 then a1 elif C2 then a2 else a0,
+                #           if C1 then b1 elif C2 then b2 else b0)"
+                # Note that no test is a function of var.
+                # First rearrange child expressions
+                cases = [p[0] for p in expr.args]
+
+                cases_ghs = [rearrange_expr(c, var) for c in cases]
+                # Now construct the new expression
+                conds = [e[1] for e in expr.args]
+
+                def piecewise_branch(i):
+                    pieces_i = zip(map(lambda gh: gh[i], cases_ghs), conds)
+                    pieces_i = [p for p in pieces_i if p[0] is not None]  # Remove cases that are None
+                    new_expr = None
+                    if pieces_i:
+                        new_expr = sp.Piecewise(*pieces_i)
+                    return new_expr
+                gh = (piecewise_branch(0), piecewise_branch(1))
+
+            else:
+                h = sp.Wild('h', exclude=[var])
+                g = sp.Wild('g', exclude=[var])
+                match = expr.expand().match(g + h * var)
+                gh = (None, None)
+                if match is not None:
+                    gh = (match[g], match[h])
+            return gh
+
+        def print_rearrange_expr(expr, var):
+            expr = sp.piecewise_fold(expr)
+            gh = rearrange_expr(expr, var)
+            return {'state_var_index': self._state_vars.index(var),
+                    'var': self._printer.doprint(var),
+                    'g': self._printer.doprint(gh[0] if gh is not None and gh[0] is not None else 0.0),
+                    'h': self._printer.doprint(gh[1] if gh is not None and gh[1] is not None else 0.0)}
+
+        # get linear state vars
+        linear_sv = [d for d in self._y_derivatives if d.args[0] not in self._non_linear_state_vars
+                     and not d.args[0] == self._membrane_voltage_var]
+
+        # fill in the equations containing state vars only
```

Comment:
I think it'd help future readers to expand on your comments explaining what this is doing!